### ISSUE TITLE
YJIT: reuse blocks for deferred compilation when possible

### ISSUE THREAD

**Issue by maximecb** [Shopify/ruby#466]
We use the `defer_compilation()` mechanism in many places in YJIT to allow us to peek at run-time values and specialize code (run-time value promotion). However, this has a memory usage cost, as we must generate a stub, and we create a new block and context object each time.

In `railsbench`, about 31% of the blocks seem to come from deferred compilation:

```
compiled_iseq_count:         2686
compiled_block_count:       25202
defer_count:                 7903
```

I think that, since we are essentially just using the deferred compilation mechanism to pause compilation, we should be able, in most cases, to avoid creating a new block and just resume/continue mutating the existing block. This may not always be possible in situations where we have green threads or ractors, because someone else could come and generate code first. However, this should be fairly rare in practice, and we should be able to detect if any code has been generated after the jump to stub or not.

If there are ractors, we might not be able to safely free the branch object that jumped to the stub in the first place, but at least we'd get to reuse the block object.

@XrXr @k0kubun do you think that this idea is viable, or can you see weird corner cases that I'm not thinking of?

I'm also wondering if it would be possible (and safe?) to overwrite the machine code for these blocks in some cases to reduce the frequency of code GC. This would involve "rewinding" the write position. This is much less important than reusing the `Block` objects themselves though.

**Comment by XrXr** [comment]
I'm not sure how this would work. For example for method calls, which is probably the most common reason we defer, we need the exact class of the receiver to lookup the method, setup assumptions etc. We can never get that info from the context because we never store it. Another major source of defer are constants, and we need to lift from the interpreter's cache for those.

For cases where this could work, it seems like it'd involve a bit of refactoring. We need to check upfront the type info each codegen method we need to decide whether we need to defer.

**Comment by maximecb** [comment]
@XrXr I'm not saying don't defer. I'm saying when we do the defer and we hit the stub, we could avoid allocating a new `Block` object and instead mutate/reuse the `Block` object that was ended due to `defer_compilation`. This may not be possible in every case, but should be possible in most cases.

**Comment by XrXr** [comment]
Oh, sorry about the misunderstanding. I think to do this safely that we'd have to rework invalidation. Right now it works on block boundary. If we put two chunks of code into the same block and the first chunk invalidates the second chunk, we'd end up running the second chunk incorrectly.

**Comment by k0kubun** [comment]
> do you think that this idea is viable, or can you see weird corner cases that I'm not thinking of?

I think this overall direction is a good first step and achievable. If a context is needed only once, we shouldn't need to keep it forever.

> I think to do this safely that we'd have to rework invalidation. Right now it works on block boundary. If we put two chunks of code into the same block and the first chunk invalidates the second chunk, we'd end up running the second chunk incorrectly.

Let's rework it. Essentially, we need two invalidation targets and a single context there, and we have two invalidation targets and two contexts today.

> I'm also wondering if it would be possible (and safe?) to overwrite the machine code for these blocks in some cases to reduce the frequency of code GC. This would involve "rewinding" the write position.

Without Ractors, if the block doesn't have `jit_prepare_routine_call` or `gen_check_ints`, it probably means that all other threads/fibers are not staying in that block, so maybe we could do that when the block is at the end of the inline block. With Ractors or interruptible code, we'd need to scan the stack to avoid rewriting on-stack blocks, so we'd need to be careful about the scanning cost.

We actually don't hit Code GC very often, so this feels like a low priority as of today (assuming the code size/locality impact is not too significant) though.

**Comment by XrXr** [comment]
> Let's rework it. Essentially, we need two invalidation targets and a single context there, and we have two invalidation targets and two contexts today.

Exits are context dependent, so each invalidation target needs their own context. We'd still save some memory, though.

**Comment by k0kubun** [comment]
Is there anything useful other than `sp_offset` (and `stack_size`?) on invalidation? Maybe we could separate invalidation contexts from optimization contexts.

**Comment by maximecb** [comment]
> Is there anything useful other than `sp_offset` (and `stack_size`?) on invalidation? Maybe we could separate invalidation contexts from optimization contexts.

You make a good point that maybe not all that information is necessary for invalidation. I've been thinking of redesigning the context objects so we can store more information, possibly using a scheme like hash consing (immutable data structure), but it might also be useful to store less information in some cases. I'd rather not introduce extra complexity into the context right away though.
